### PR TITLE
Update Parser.php

### PR DESCRIPTION
**Pull request by SerafimArts**
Fix "strrpos()" bug when offset is greater than the length of haystack string.
Fix https://github.com/hoaproject/Compiler/issues/71

### CODE REVIEW

**Comment by SerafimArts** [comment]
It would be good to show (in UnexpectedToken exception) in this case (error in line 0) not the previous token `(null)`. 

Meybe some other message, defined in `*.pp` file?
`grammar.pp`
```
%undefined T_UNDEFINED

// Unexpected token (T_UNDEFINED) at line 0 and column 0: ...
// test
// ↑
```

File: Llk/Parser.php [hunk 1]
```diff
@@ -192,7 +192,7 @@ public function parse($text, $rule = null, $tree = true)
                 $column = 1;
 
                 if (!empty($text)) {
-                    if (0 === $offset) {
+                    if (0 === $offset || null === $offset) {
```
Comment (Hywan) [inline Llk/Parser.php]: So `$offset` can be set to 0 or `null`. Is it because `$tokens` does not exist or because `$token['offset']`is really equal to `null`?

**Comment by Hywan** [comment]
Thank @Pierozi for handling this PR and for the review!

File: Llk/Parser.php (hunk 1 above)
Comment (SerafimArts) [inline Llk/Parser.php]: @Hywan offset really equal to null. The problem can be reproduced if the first token and/or character does not correspond to any rule (rule == defined token).

**Comment by SerafimArts** [comment]
@Hywan The problem can be reproduced if the first token and/or character does not correspond to any rule (rule == defined token).

**Comment by Hywan** [comment]
@SerafimArts I took the liberty to implement another approach, please see https://github.com/hoaproject/Compiler/commit/23cc8e191bdd9dd3aec73c7b450200ebccc0deb7. The patch is very minimal too.

You are right, an error token is not always found when backtracking, especially when it happens at the very first position of the text being parsed. In this particular edge case, the current token is used as the error token. So with your example, it produces this error:

```
Hoa\Compiler\Llk\Parser::parse(): (0) Unexpected token "type" (T_NAME) at line 1 and column 1:
type Test { }
↑
```

instead of:

```
Hoa\Compiler\Llk\Parser::parse(): (0) Unexpected token "(null)" ((null)) at line 1 and column 1:
type Test { }
↑
```

as your patch suggested.

Thank you very much for this excellent bug report and your patience!

**Comment by Hywan** [comment]
[3.17.08.08 has been released](https://github.com/hoaproject/Compiler/releases/tag/3.17.08.08) because this is an important bug fix.

**Comment by SerafimArts** [comment]
@Hywan i think this commit https://github.com/hoaproject/Compiler/commit/23cc8e191bdd9dd3aec73c7b450200ebccc0deb7 are better. I also thought about it, but decided to limit myself to simple bugfix =)

**Comment by Hywan** [comment]
:+1: The harder part was to find the bug, and what you did was excellent. Thank you!
